dev-tools/releng/src/job.rs

```diff
@@ -0,0 +1,284 @@
+        logger: &Logger,
+        permits: Arc<Semaphore>,
+        log_dir: &Utf8Path,
+    ) -> Jobs {
+        Jobs {
+            logger: logger.clone(),
+            permits,
+            log_dir: log_dir.to_owned(),
+            map: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn push<F>(
+        &mut self,
+        name: impl AsRef<str>,
+        future: F,
+    ) -> Selector<'_>
+    where
+        F: Future<Output = Result<()>> + 'static,
+    {
+        let name = name.as_ref().to_owned();
+        assert!(!self.map.contains_key(&name), "duplicate job name {}", name);
+        self.map.insert(
+            name.clone(),
+            Job {
+                future: Box::pin(run_job(
+                    self.logger.clone(),
+                    self.permits.clone(),
+                    name.clone(),
+                    future,
+                )),
+                wait_for: Vec::new(),
+                notify: Vec::new(),
+            },
+        );
+        Selector { jobs: self, name }
+    }
+
+    pub(crate) fn push_command(
+        &mut self,
+        name: impl AsRef<str>,
+        command: &mut Command,
+    ) -> Selector<'_> {
+        let name = name.as_ref().to_owned();
+        assert!(!self.map.contains_key(&name), "duplicate job name {}", name);
+        self.map.insert(
+            name.clone(),
+            Job {
+                future: Box::pin(spawn_with_output(
+                    // terrible hack to deal with the `Command` builder
+                    // returning &mut
+                    std::mem::replace(command, Command::new("false")),
+                    self.logger.clone(),
+                    self.permits.clone(),
+                    name.clone(),
+                    self.log_dir.join(&name).with_extension("log"),
+                )),
+                wait_for: Vec::new(),
+                notify: Vec::new(),
+            },
+        );
+        Selector { jobs: self, name }
+    }
+
+    pub(crate) fn select(&mut self, name: impl AsRef<str>) -> Selector<'_> {
+        Selector { jobs: self, name: name.as_ref().to_owned() }
+    }
+
+    pub(crate) async fn run_all(self) -> Result<()> {
+        self.map
+            .into_values()
+            .map(Job::run)
+            .collect::<FuturesUnordered<_>>()
+            .try_collect::<()>()
+            .await
+    }
+}
+
+impl Job {
+    async fn run(self) -> Result<()> {
+        let result: Result<(), RecvError> = self
+            .wait_for
+            .into_iter()
+            .collect::<FuturesUnordered<_>>()
+            .try_collect::<()>()
+            .await;
+        result.map_err(|_| anyhow!("dependency failed"))?;
+
+        self.future.await?;
+        for sender in self.notify {
+            sender.send(()).ok();
+        }
+        Ok(())
+    }
+}
+
+impl<'a> Selector<'a> {
+    #[track_caller]
+    pub(crate) fn after(self, other: impl AsRef<str>) -> Self {
+        let (sender, receiver) = oneshot::channel();
```

Comment:
I've used this pattern before, and I love how it makes respecting the DAG Tokio's responsibility. Can that be made explicit in a comment for those unfamiliar with it?